File: android/src/main/kotlin/com/exponea/data/InAppMessageAction.kt
package com.exponea.data

import com.exponea.sdk.models.InAppMessage
import com.exponea.sdk.models.InAppMessageButton
import com.google.gson.Gson

data class InAppMessageAction(
    val type: InAppMessageActionType,
    val message: InAppMessage?,
    val button: InAppMessageButton? = null,
    val interaction: Boolean? = null,
    val errorMessage: String? = null,
) {
    fun toMap(): Map<String, Any?> {
        return mapOf(
            "type" to type.value,
            "message" to message?.let { InAppMessageCoder.encode(it) },
            "button" to button?.let {
                mapOf(
                    "text" to it.text,
                    "url" to it.url
                )
            },
            "interaction" to interaction,
            "errorMessage" to errorMessage
        ).filterValues { it != null }
    }
}

File: ios/Classes/Data/InAppMessageAction.swift
//
//  InAppMessageAction.swift
//  exponea
//

import Foundation
import ExponeaSDK

final class InAppMessageAction {
    let type: InAppMessageActionType
    let message: InAppMessage?
    let button: InAppMessageButton?
    let interaction: Bool?
    let errorMessage: String?
    
    init(type: InAppMessageActionType, message: InAppMessage? = nil, button: InAppMessageButton? = nil, interaction: Bool? = nil, errorMessage: String? = nil) {
        self.type = type
        self.message = message
        self.button = button
        self.interaction = interaction
        self.errorMessage = errorMessage
    }
    
    func toMap() -> [String:Any?]  {
        var data : [String:Any?] = [
            "type": String(describing: type),
        ]
        if let message {
            data["message"] = try? InAppMessageCoder.encode(message)
        }
        if let button {
            data["button"] = [
                "text": button.text,
                "url": button.url
            ]
        }
        if let interaction {
            data["interaction"] = interaction
        }
        if let errorMessage {
            data["errorMessage"] = errorMessage
        }
        return data
    }
}

public enum InAppMessageActionType {
    case click
    case close
    case error
    case show
}
